pyiron_contrib/workflow/node.py

```diff
@@ -364,12 +364,17 @@ def __init__(
         if update_on_instantiation:
             self.update()
 
+    @property
+    def _input_args(self):
+        return inspect.signature(self.node_function).parameters
+
     def _build_input_channels(self, storage_priority: dict[str:int]):
         channels = []
         type_hints = get_type_hints(self.node_function)
-        parameters = inspect.signature(self.node_function).parameters
 
-        for label, value in parameters.items():
+        for label, value in self._input_args.items():
+            if label == "self":
```

Comment:
I was wondering whether to check it as well, but I decided not to, because in practice, when I'm running the code locally, I might not want to use `self`, i.e. I can imagine a case like this:


```python
def run_simulation(x, job, self=None):
    if self is not None:
        job.server.cores = 4
    job.input.x = x
    job.run()
    return job.output.y
```

Then I probably don't want to have to set `self` every time, which would be required if it's the first argument.

And whether it's a social scientist or not, they should be able to know the minimum of python, and would avoid `self`.

Reply:
Ah, what I had in mind was that when `self` is in the function definition, the user would _never_ need to set its value -- when converting from function to class, the class would ensure that the class instance was _always_ passed in to this value.

Thus, in  your example `if self is not None` wouldn't make sense, but this would:

```python
def run_simulation(self, x, job):
    if self.server is not None:
        job.server = self.server  # or all the same but with server.cores maybe
    job.input.x  = x
    job.run()
    return job.output.y
```

Then you'd decorate this with `@node` and do stuff like 

```python
my_run_node = run_simulation(x=42, job=an_instance_defined_earlier)
# or set with my_run_node.x = 42, etc.
# Calling run_simulation(self=foo) or my_run_node.self = foowill _not work_ 
# because it does not match a parsed function kwarg!
my_run_node.server = Server("cm")  # or whatever/however we attach one if not there by default
my_run_node.server.cores = 4
my_run_node.run()
```

(again, with the disclaimer that at some point we want to bring `Server` into line with the `Executor` stuff Jan has been working on recently, but this is still the gist of it)

> And whether it's a social scientist or not, they should be able to know the minimum of python, and would avoid `self`.

I kind of agree...I just thought that if they're ~~dumb~~ uninformed enough to not be putting `self` first, then they're probably uninformed enough to be trying to use it as a regular variable.

Reply:
> Ah, what I had in mind was that when self is in the function definition, the user would never need to set its value -- when converting from function to class, the class would ensure that the class instance was always passed in to this value.

yes if it’s used in the frame of workflow I agree, but I’m more talking about the case that the function is used independently.

Reply:
I changed it. Now if `self` is used but not in the first position, it is treated as a normal argument, but pyiron issues a warning, which I think is fair because a rational programmer should never use `self` as a normal argument in python.

Reply:
@samwaseda Good idea. I think this hits exactly the right balance now.